src/design-pages/passkey_kerberos.rst

```diff
@@ -0,0 +1,262 @@
+
+Related ticket(s):
+------------------
+ * https://github.com/SSSD/sssd/issues/6228
+
+Problem statement
+-----------------
+The use of new tools to authenticate users, such as 2FA, U2F and FIDO2, is
+becoming increasingly popular. Currently SSSD provides local authentication of
+a centrally managed user with passkeys, but it doesn't provide any way to
+authenticate remotely to a system. This diminishes the value provided by SSSD
+in some environments like big companies and governments, where remote
+authentication is a common pattern.
+
+This design is an extension of the
+`enable passkey authentication <https://sssd.io/design-pages/passkey_authentication.html>`__,
+where a centrally managed user can authenticate locally in a system with a
+passkey token. It describes a Kerberos pre-authentication mechanism using
+passkey authentication. Once it is completed, the received initial ticket
+granting ticket (TGT) can be used for single sign-on into other
+Kerberos-enabled services.
+
+For that purpose SSSD needs to be modified to accommodate the new workflow,
+while still allowing the old one, where the user was authenticated locally
+without issuing any Kerberos ticket. The new workflow is used for online
+authentication against an IPA server. The old workflow is used for any offline
+authentication and for AD and other LDAP servers.
+
+For the purpose of this feature, passkey is a FIDO2 compatible device supported
+by the libfido2 library.
+
+Use cases
+---------
+As a user, not only do I want to sit in front of a system enrolled into a
+centralized identity management solution and log in (desktop and/or terminal)
+using a passkey device connected to the system, but I also want to get a
+Kerberos ticket to identify myself to other services.
+
+Solution overview
+-----------------
+.. uml:: ../diagrams/passkey_kerberos_architecture_diagram.pu
+
+* SSSD provides a pre-authentication method implementation (pre-auth plugin) to
+  MIT Kerberos library.
+
+* When both the libkrb5 and the KDC support the same pre-authentication method,
+  KDC responds to an AS-REQ message with a request to provide the assertion
+  data required by the pre-authentication method.
+  
+* SSSD's pre-auth plugin obtains the assertion data from the passkey.
+
+* The assertion data is sent by libkrb5 back to the KDC server as a part of the
+  AS-REQ exchange.
+
+* The KDC server in turn relays the request to the ipa-otpd, which retrieves
+  the user information and ask the passkey_child to validate the assertion.
+
+* Once validated the Kerberos KDC issues a ticket and sends it to SSSD via
+  libkrb5.
+
+Authentication method policy
+****************************
+
+This design adds some complexity to the passkey authentication, and this may
+lead to a degraded user experience. For this reason, the following
+authentication policy is implemented:
+
+* In the IPA case, if the server is online and a passkey entry is found in the
```

Comment:
Hi,

I think in the online case a "policy" "try only methods which will return a Kerberos ticket" makes sense. But even "if online and a passkey entry is found" this might not be possible wither because the IPA server does not has the Kerberos plugin installed or the user entry does not have `passkey` listed in the user-auth-types. In those cases I wonder if having passkey authentication available by default for offline authentication might irritate the users?

So it might be worth to store the authentication methods which are available online in the user object to have a consistent experience when doing offline authentication?

Additionally, to make the passkey usage more flexible, it might be worth to add an option for the backends to allow passkey authentication although not supported via Kerberos on the server side. This might be useful for the AD use-case as well to allow passkey authentication for users which have the passkey data set in the user entry. Since this would be an exception to the  "try only methods which will return a Kerberos ticket" "policy" this option should not be enable by default. But in environments which Kerberos tickets are not that important it would help to support the usage of local passkey authentication. If the passkey prompt should come first in those cases and can be skipped if users wants to have a login method which returns a Kerberos ticket or if it should come last after method which can get a Kerberos ticket might then be something which should be configurable in the PAM responder to have a consistent behavior if multiple backends are configured.

For `auth_provider = ldap` this option can be enabled by default, i.e. if the user has passkey data configured local passkey authentication is used and if not authentication against the LDAP server with an authenticated LDAP bind is used.

I wonder if this option can be boolean of if there might be other use-cases like not allowing passkey when online but only when offline? Or only allowing passkey authentication (this would look like an `auth_provider = passkey` but since we support Kerberos with passkey in IPA I think it would be better to have this as an option for all the existing providers).

About skipping passkey authentication. I think it would be good to allow this either if the user enters something at the interactive prompt or nothing at the pin prompt. This would cover the case where a pin is not needed but the interactive prompt might be extended with  e.g. "... or press 's' and then enter to skip passkey authentication".

bye,
Sumit

Reply:
given that this policy should apply to all passwordless methods, including smartcards and IdP (in IPA case), may be it would be best to have a bit less specific set of options? 

E.g. something like  `local_auth_policy = match | disable | enable` where `match` is to match offline and online state for Kerberos methods, `disable` would disallow those methods that do not lead to Kerberos ticket, and `enable` would allow any local methods.

Reply:
Hi,

about `passkey_auth_priority`, I think this functionality should be part of a more generic `order` option for the ´[prompting]` section. If you set `passkey` first with this option, e.g. `order = passkey, 2fa, password`, then passkey authentication would be tried first if available for the user.

About `allow_local_passkey`, I agree with Alexander that this option should be more generic to cover other authentication methods as well. However, I think it would be good to have a way to only enable/allow specific method, so maybe options like `enable:passkey` to only allow passkey for local authentication. This can be handled with the `order` option as well, e.g. `order = passkey, password` would only allow passkey authentication and no other local methods like e.g. Smartcard. But since this will control the behavior for all configured domains it mightr be better to have something on the domain level as well. E.g. if there are two domains configured and one domain should allow local passkey authentication while the other domains local Smartcard authentication.

Additionally, a fourth option `only`  for ` local_auth_policy` which would ignore the online methods and only offer the local ones might be useful. This would cover e.g. the AD use case where only passkey authentication should be used.

bye,
Sumit